crates/matrix-sdk/src/test_utils/mocks/mod.rs

```diff
@@ -3243,6 +3243,244 @@ impl<'a> MockEndpoint<'a, LoginEndpoint> {
+        self 
+    }
+    
+    /// sets expires_in_ms
+    pub fn set_expires_in_ms(&mut self, value: u32) -> &mut Self { 
+        self.expires_in_ms = Some(value); 
+        self 
+    }
+    
+    /// sets refresh_token
+    pub fn set_refresh_token(&mut self, value: &str) -> &mut Self { 
+        self.refresh_token = Some(value.to_string()); 
+        self 
+    }
+    
+    /// sets user_id
+    pub fn set_user_id(&mut self, value: &str) -> &mut Self { 
+        self.user_id = Some(value.to_string()); 
+        self 
+    }
+    
+    /// sets homeserver_url
+    pub fn set_homeserver_url(&mut self, value: &str) -> &mut Self { 
+        self.homeserver_url = Some(value.to_string()); 
+        self 
+    }
+    
+    /// sets identity_url
+    pub fn set_identity_url(&mut self, value: &str) -> &mut Self { 
+        self.identity_url = Some(value.to_string()); 
+        self 
+    }
+    
+    /// builds a ResponseTemplate from builder
+    pub fn build(&mut self) -> ResponseTemplate {
+        if self.access_token.is_none() {
+            panic!("access token not given");
+        } 
+        if self.device_id.is_none() {
+            panic!("device id not given");
+        } 
+        if self.user_id.is_none() {
+            panic!("user id not given");            
+        } 
+
+        if self.homeserver_url.is_none() && self.identity_url.is_some() {
+            panic!("identity uri given but not homeserver uri. homeserver uri required for well known field.")
+        }
+
+        let mut body = Map::new();
+        body.insert("access_token".to_string(), Value::String(self.access_token.clone().unwrap()));
+        body.insert("device_id".to_string(), Value::String(self.device_id.clone().unwrap()));
+        body.insert("user_id".to_string(), Value::String(self.user_id.clone().unwrap()));
+        
+        if self.expires_in_ms.is_some() {
+            body.insert("expires_in_ms".to_string(), Value::Number(Number::from(self.expires_in_ms.clone().unwrap())));
+        }
+        
+        if self.refresh_token.is_some() {
+            body.insert("refresh_token".to_string(), Value::String(self.refresh_token.clone().unwrap()));
+        }
+        
+        if self.homeserver_url.is_some() {
+            let mut well_known = Map::new();
+            let mut home_server = Map::new();
+            home_server.insert("base_url".to_string(), Value::String(self.homeserver_url.clone().unwrap()));
+            well_known.insert("m.homeserver".to_string(), Value::Object(home_server));
+
+            if self.identity_url.is_some() {
+                let mut identity_server = Map::new();
+                identity_server.insert("base_url".to_string(), Value::String(self.identity_url.clone().unwrap()));
+                well_known.insert("m.identityserver".to_string(), Value::Object(identity_server));
+            }
+
+            body.insert("well_known".to_string(), Value::Object(well_known));
+        }
+        
```

Comment:
This is a very manual way of building the `json::Value` from the `ResponseTemplate`.

You could have just used `json!({})` like the `RoomRelationsResponseTemplate` does:

https://github.com/matrix-org/matrix-rust-sdk/blob/72d133260c29e132a75597ae6ca77963813e74fa/crates/matrix-sdk/src/test_utils/mocks/mod.rs#L3067-L3072

There is no difference between an unset field and a field being set to `None`, so you don't need to make this conditional at all.